modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,336 @@
+package registry
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/prometheus/model/exemplar"
+	promhistogram "github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type nativeHistogram struct {
+	metricName string
+
+	// TODO we can also switch to a HistrogramVec and let prometheus handle the labels. This would remove the series map
+	//  and all locking around it.
+	//  Downside: you need to list labels at creation time while our interfaces only pass labels at observe time, this
+	//  will requires a bigger refactor, maybe something for a second pass?
+	//  Might break processors that have variable amount of labels...
+	//  promHistogram prometheus.HistogramVec
+
+	seriesMtx sync.Mutex
+	series    map[uint64]*nativeHistogramSeries
+
+	onAddSerie    func(count uint32) bool
+	onRemoveSerie func(count uint32)
+
+	buckets []float64
+
+	traceIDLabelName string
+}
+
+type nativeHistogramSeries struct {
+	// labels should not be modified after creation
+	labels        LabelPair
+	promHistogram prometheus.Histogram
+	lastUpdated   *atomic.Int64
```

Comment:
does this need to be atomic? it seems like its always protected by the seriesMtx above

Reply:
It appears that we are always under lock as you mention.  Do you suggest that `int64` might be a better type here?

Reply:
yup, if it's already protected by the global lock, would prefer an `int64`

Reply:
A historical note since this code is copied from the original histogram implementation: we used to protect the series map with a `RWMutex`, this way we could grab the serie and update the atomic fields individually without holding a lock. After finding some concurrency issues updating the counters while reading them we opted to lock the entire map with a normal mutex.

Since `prometheus.Histogram` is concurrency-safe we could opt to switch back to a `RWMutex`. But I think this area might still change a lot has we integrate the prometheus code better, so it's fine to keep as-is for now imo.